src/Bundle/spec/Doctrine/DBAL/DataSourceSpec.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace spec\Sylius\Bundle\GridBundle\Doctrine\DBAL;
+
+use Doctrine\DBAL\Query;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Pagerfanta\Pagerfanta;
+use PhpSpec\ObjectBehavior;
+use Sylius\Component\Grid\Data\DataSourceInterface;
+use Sylius\Component\Grid\Parameters;
+
+final class DataSourceSpec extends ObjectBehavior
+{
+    function it_implements_data_source(QueryBuilder $queryBuilder): void
+    {
+        $this->beConstructedWith($queryBuilder);
+
+        $this->shouldImplement(DataSourceInterface::class);
+    }
+
+    function it_gets_the_data(QueryBuilder $queryBuilder): void
+    {
+        $queryBuilder->getType()->willReturn(QueryBuilder::SELECT);
+        $this->beConstructedWith($queryBuilder);
+
+        $this->getData(new Parameters(['page' => '1']))->shouldHaveType(Pagerfanta::class);
```

Comment:
To be honest, it does not check too much :( Probably the best option would be to create a [custom matcher](https://github.com/Sylius/SyliusResourceBundle/blob/0bb5ab5dd36edc2e6402d22f0431f2d600efe488/src/Component/spec/Generator/RandomnessGeneratorSpec.php#L52) that would check if the returned object:

- is type of `Pagerfanta`
- has `$normalizeOutOfRangePages` set properly
- has `$currentPage` set as it should be

Then we're quite safe from regression :)

Reply:
done, I did not use the custom matcher, but I tried to and was not be satisfied.